sw/device/silicon_creator/lib/otbn_util_unittest.cc

```diff
@@ -0,0 +1,176 @@
+                       {
+                           {OTBN_INTR_COMMON_DONE_BIT, 1},
+                       });
+
+    EXPECT_ABS_READ32(base_ + OTBN_ERR_BITS_REG_OFFSET, err_bits);
+    EXPECT_ABS_READ32(base_ + OTBN_STATUS_REG_OFFSET, kOtbnStatusIdle);
+
+    if (err_bits == kOtbnErrBitsNoError) {
+      EXPECT_ABS_READ32(base_ + OTBN_STATUS_REG_OFFSET, kOtbnStatusIdle);
+    }
+  }
+
+  uint32_t base_ = TOP_EARLGREY_OTBN_BASE_ADDR;
+  rom_test::MockAbsMmio abs_mmio_;
+  rom_test::MockRnd rnd_;
+  rom_test::MockSecMmio sec_mmio_;
+};
+
+class OtbnAppTest : public OtbnTest {};
+
+TEST_F(OtbnAppTest, OtbnLoadAppSuccess) {
+  std::array<uint32_t, 2> imem_data = {0x01234567, 0x89abcdef};
+  std::array<uint32_t, 2> dmem_data = {0x456789ab, 0xcdef0123};
+  otbn_app_t app = {
+      .imem_start = imem_data.data(),
+      .imem_end = imem_data.data() + imem_data.size(),
+      .dmem_data_start = dmem_data.data(),
+      .dmem_data_end = dmem_data.data() + imem_data.size(),
+  };
+
+  // Test assumption.
+  static_assert(OTBN_DMEM_SIZE_BYTES >= 4 * dmem_data.size(),
+                "OTBN DMEM size too small");
+  static_assert(OTBN_IMEM_SIZE_BYTES >= 4 * imem_data.size(),
+                "OTBN IMEM size too small");
+
+  otbn_t otbn;
+  otbn_init(&otbn);
+
+  // `otbn_busy_wait_for_done` - read twice due to hardening.
+  EXPECT_ABS_READ32(base_ + OTBN_STATUS_REG_OFFSET, kOtbnStatusIdle);
+  EXPECT_ABS_READ32(base_ + OTBN_STATUS_REG_OFFSET, kOtbnStatusIdle);
+  // `otbn_imem_sec_wipe`
+  ExpectCmdRun(kOtbnCmdSecWipeImem, kOtbnErrBitsNoError);
+  // `otbn_imem_write`
+  EXPECT_CALL(rnd_, Uint32()).WillOnce(Return(0));
+  EXPECT_ABS_WRITE32(base_ + OTBN_IMEM_REG_OFFSET, imem_data[0]);
+  EXPECT_ABS_WRITE32(base_ + OTBN_IMEM_REG_OFFSET + 4, imem_data[1]);
+  // `otbn_dmem_sec_wipe`
+  ExpectCmdRun(kOtbnCmdSecWipeDmem, kOtbnErrBitsNoError);
+  // `otbn_dmem_write`
+  EXPECT_CALL(rnd_, Uint32()).WillOnce(Return(0));
+  EXPECT_ABS_WRITE32(base_ + OTBN_DMEM_REG_OFFSET, dmem_data[0]);
+  EXPECT_ABS_WRITE32(base_ + OTBN_DMEM_REG_OFFSET + 4, dmem_data[1]);
+
+  EXPECT_EQ(otbn_load_app(&otbn, app), kErrorOk);
+}
+
+TEST_F(OtbnAppTest, OtbnLoadInvalidApp) {
+  // Create an invalid app with an empty IMEM range.
+  std::array<uint32_t, 0> imem_data = {};
+  std::array<uint32_t, 2> dmem_data = {0x456789ab, 0xcdef0123};
+  otbn_app_t app = {
+      .imem_start = imem_data.data(),
+      .imem_end = imem_data.data() + imem_data.size(),
+      .dmem_data_start = dmem_data.data(),
+      .dmem_data_end = dmem_data.data() + dmem_data.size(),
+  };
+
+  otbn_t otbn;
+  otbn_init(&otbn);
+
+  EXPECT_EQ(otbn_load_app(&otbn, app), kErrorOtbnInvalidArgument);
+}
+
+TEST_F(OtbnAppTest, OtbnExecuteApp) {
+  EXPECT_SEC_WRITE32(base_ + OTBN_CTRL_REG_OFFSET, 0x01);
+  ExpectCmdRun(kOtbnCmdExecute, kOtbnErrBitsNoError);
+
+  EXPECT_EQ(otbn_execute(), kErrorOk);
+}
+
+class OtbnWriteTest : public OtbnTest {};
+
+TEST_F(OtbnWriteTest, Success) {
+  const uint32_t dest = 6;
```

Comment:
Prefer `constexpr` in C++, [see](https://docs.opentitan.org/doc/sg/c_cpp_coding_style/#features).
Constants should be named like `kDest`, [see](https://docs.opentitan.org/doc/sg/c_cpp_coding_style/#naming).
You can also use a unamed enum here, [see](https://docs.opentitan.org/doc/sg/c_cpp_coding_style/#constants-and-preprocessor-macros).